File: main/server/vial/auth_manager.py
import jwt,hashlib,logging
from datetime import datetime,timedelta
from fastapi import HTTPException

logger=logging.getLogger(__name__)

class AuthManager:
    """Manages authentication with API keys and JWT tokens, tied to wallet IDs."""
    def __init__(self):
        """Initialize AuthManager with secret key and API key mappings."""
        self.secret_key="vial-mcp-secret-2025"
        self.api_keys={
            "api-a24cb96b-96cd-488d-a013-91cb8edbbe68":{"user_id":"vial_user","wallet_id":"wallet_123","created_at":datetime.now()},
            "api-bd9d62ec-a074-4548-8c83-fb054715a870":{"user_id":"vial_user","wallet_id":"wallet_456","created_at":datetime.now()}
        }

    def verify_api_key(self,api_key:str)->bool:
        """Verify if the API key is valid and associated with a wallet.

        Args:
            api_key (str): API key to verify.

        Returns:
            bool: True if valid, False otherwise.
        """
        try:
            if api_key not in self.api_keys:
                logger.warning(f"Invalid API key attempted: {api_key}")
                return False
            return True
        except Exception as e:
            logger.error(f"API key verification failed: {str(e)}")
            return False

    def generate_token(self,api_key:str)->str:
        """Generate a JWT token for a valid API key.

        Args:
            api_key (str): API key to generate token for.

        Returns:
            str: JWT token.

        Raises:
            HTTPException: If token generation fails.
        """
        try:
            if not self.verify_api_key(api_key):
                raise HTTPException(status_code=401,detail="Invalid API key")
            payload={"user_id":self.api_keys[api_key]["user_id"],"wallet_id":self.api_keys[api_key]["wallet_id"],"exp":datetime.utcnow()+timedelta(hours=24)}
            token=jwt.encode(payload,self.secret_key,algorithm="HS256")
            logger.info(f"Generated token for wallet {self.api_keys[api_key]['wallet_id']}")
            return token
        except Exception as e:
            logger.error(f"Token generation failed: {str(e)}")
            raise HTTPException(status_code=500,detail=f"Token generation failed: {str(e)}")

    def verify_token(self,token:str)->dict:
        """Verify a JWT token and return its payload.

        Args:
            token (str): JWT token to verify.

        Returns:
            dict: Decoded token payload.

        Raises:
            HTTPException: If token is invalid or expired.
        """
        try:
            payload=jwt.decode(token,self.secret_key,algorithms=["HS256"])
            logger.info(f"Token verified for wallet {payload['wallet_id']}")
            return payload
        except jwt.ExpiredSignatureError:
            logger.warning("Token expired")
            raise HTTPException(status_code=401,detail="Token expired")
        except jwt.InvalidTokenError:
            logger.warning("Invalid token")
            raise HTTPException(status_code=401,detail="Invalid token")
        except Exception as e:
            logger.error(f"Token verification failed: {str(e)}")
            raise HTTPException(status_code=500,detail=f"Token verification failed: {str(e)}")

File: main/api/mcp/lib/auth_manager.js
const jwt = require('jsonwebtoken');
const logger = require('./logger');

class AuthManager {
  constructor() {
    this.secret = process.env.JWT_SECRET || 'default-secret';
  }

  async verifyApiKey(apiKey) {
    // In production, this would validate against a stored key or external service
    // For demo, accept any non-empty key
    if (!apiKey || apiKey.length < 10) {
      logger.error('Invalid API key');
      return false;
    }
    logger.info(`API key verified: ${apiKey.substring(0, 4)}...`);
    return true;
  }

  async generateToken(apiKey) {
    try {
      const payload = {
        user_id: 'a1d57580-d88b-4c90-a0f8-6f2c8511b1e4', // Mock user ID
        api_key: apiKey,
        iat: Math.floor(Date.now() / 1000),
        exp: Math.floor(Date.now() / 1000) + 86400 // 24 hours
      };
      const token = jwt.sign(payload, this.secret);
      logger.info('JWT token generated');
      return token;
    } catch (err) {
      logger.error(`Token generation error: ${err.message}`);
      throw err;
    }
  }

  async verifyToken(token) {
    try {
      const decoded = jwt.verify(token, this.secret);
      logger.info(`Token verified for user: ${decoded.user_id}`);
      return decoded;
    } catch (err) {
      logger.error(`Token verification error: ${err.message}`);
      return null;
    }
  }
}

module.exports = { AuthManager, verifyToken: async (token) => new AuthManager().verifyToken(token) };
